src/beacon.cpp

```diff
@@ -3,96 +3,69 @@
         {
-            fResult = VerifyCPIDSignature(cpid, hashBlock.GetHex(), sSignature);
-
-            if (fResult)
+            if (pwalletMain->GetKey(oldKey.GetID(), outPrivPubKey))
             {
-                LogPrintf("GenerateBeaconKeys::Current keypair is valid.");
+                assert(outPrivPubKey.IsValid());//GetKey gives valid key or false
+                assert(outPrivPubKey.GetPubKey()==oldKey);//GetKey should return the key we ask for
+                LogPrintf("GenerateBeaconKeys: Reusing >5 <6 m old Key");
                 return true;
             }
-
-            else
-                LogPrintf("GenerateBeaconKeys::Signing block with CPID was successful; However Verifying CPID Sign was not; Key pair is not valid, generating new key pair");
         }
     }
 
-    // Generate the Keypair
-    CKey key;
-    key.MakeNewKey(false);
-    CPrivKey vchPrivKey = key.GetPrivKey();
-    sOutPrivKey = HexStr<CPrivKey::iterator>(vchPrivKey.begin(), vchPrivKey.end());
-    sOutPubKey = HexStr(key.GetPubKey().Raw());
+    // Generate a new key that is added to wallet
+    CPubKey newKey;
+    if (!pwalletMain->GetKeyFromPool(newKey, false))
+        return error("GenerateBeaconKeys: Failed to get Key from Wallet");
 
-    return true;
-}
-
-bool StoreBeaconKeys(
-        const std::string &cpid,
-        const std::string &pubKey,
-        const std::string &privKey)
-{
-    if (   !WriteKey("publickey" + cpid + GetNetSuffix(), pubKey)
-        || !WriteKey("privatekey" + cpid + GetNetSuffix(), privKey)
-       )
-        return false;
+    CKeyID keyID = newKey.GetID();
 
-    else
-        return true;
-}
+    // Assign a description for this address
+    std::string strLabel= "DPoR Beacon CPID "+cpid+" "+ToString(nBestHeight);
+    pwalletMain->SetAddressBookName(keyID, strLabel);
 
-std::string GetStoredBeaconPrivateKey(const std::string& cpid)
-{
-    return GetArgument("privatekey" + cpid + GetNetSuffix(), "");
-}
+    // Get the private part of the key from wallet
+    if (!pwalletMain->GetKey(keyID, outPrivPubKey))
+        return error("GenerateBeaconKeys: Failed to get Private Key from Wallet");
 
-std::string GetStoredBeaconPublicKey(const std::string& cpid)
-{
-    return GetArgument("publickey" + cpid + GetNetSuffix(), "");
+    return true;
 }
 
-void ActivateBeaconKeys(
-        const std::string &cpid,
-        const std::string &pubKey,
-        const std::string &privKey)
+bool GetStoredBeaconPrivateKey(const std::string& cpid, CKey& outPrivPubKey)
 {
-    SetArgument("publickey" + cpid + GetNetSuffix(), pubKey);
-    SetArgument("privatekey" + cpid + GetNetSuffix(), privKey);
+    AssertLockHeld(cs_main);
+    AssertLockHeld(pwalletMain->cs_wallet);
+    CPubKey oldKey(ParseHex(GetBeaconPublicKey(cpid, false)));
+    if(oldKey.IsValid())
+    {
+        if (pwalletMain->GetKey(oldKey.GetID(), outPrivPubKey))
+        {
+            return outPrivPubKey.IsValid();
+        }
+    }
+    return false;
```

Comment:
We need to fall back on config stored keys here for when the import hasn't taken place yet.